File: javascript/quicksort.js
function quicksort(arr, partitionFn) {
    if (arr.length > 1) {
        const p = partitionFn(arr);
        return [...quicksort(arr.slice(0, p), partitionFn), ...quicksort(arr.slice(p), partitionFn)];
    }
    return arr;
}

function partitionLeft(arr) {
    const pivotIdx = 0;
    const pivot = arr[pivotIdx];

    let i = arr.length - 1;
    for (let j = i; j > 0; j--) {
        if (arr[j] > pivot) {
            const aux = arr[j];
            arr[j] = arr[i];
            arr[i] = aux;
            i--;
        }
    }

    const aux = arr[i];
    arr[i] = pivot;
    arr[pivotIdx] = aux;

    return i + 1;
}

function partitionRight(arr) {
    const pivotIdx = arr.length - 1;
    const pivot = arr[pivotIdx];

    let i = 0;
    arr.forEach((el, idx) => {
        if (el < pivot) {
            arr[idx] = arr[i];
            arr[i] = el;
            i++;
        }
    });

    const aux = arr[i];
    arr[i] = pivot;
    arr[pivotIdx] = aux;

    return i;
}

function partition(arr) {
    let [left, right] = [0, arr.length - 1];
    const pivot = arr[Math.floor(arr.length / 2)];

    while (1) {
        while (arr[left] < pivot) {
            left++;
        }
        while (arr[right] > pivot) {
            right--;
        }

        if (left >= right) {
            break;
        }

        const aux = arr[left];
        arr[left] = arr[right];
        arr[right] = aux;
    }

    return right;
}

const array = [2,1,6,4,5,3];
console.log("Original", array);
console.log("Sorted (Left)", quicksort([...array], partitionLeft));
console.log("Sorted (Right)", quicksort([...array], partitionRight));
console.log("Sorted (Middle)", quicksort([...array], partition));

File: go/quicksort.go
package main

import "fmt"

func main() {
	list := []int{5, 1, 4, 6, 3, 2}
	aux := make([]int, len(list))

	copy(aux, list)
	println("Quicksort (Right Most)")
	fmt.Println(aux)
	quickSort2(aux, partitionRightMost)
	fmt.Println(aux)

	copy(aux, list)
	println("Quicksort (Left Most)")
	fmt.Println(aux)
	quickSort2(aux, partitionLeftMost)
	fmt.Println(aux)

	copy(aux, list)
	println("Quicksort (Middle)")
	fmt.Println(aux)
	quickSort(aux, 0, len(aux)-1)
	fmt.Println(aux)
}

func quickSort(list []int, left int, right int) {
	if left >= 0 && left < right {
		p := partition(list, left, right)
		quickSort(list, left, p)
		quickSort(list, p+1, right)
	}
}

func partition(list []int, left int, right int) int {
	pivot := list[(right+left)/2]

	for {
		for list[left] < pivot {
			left += 1
		}
		for list[right] > pivot {
			right -= 1
		}

		if left >= right {
			return right
		}

		aux := list[right]
		list[right] = list[left]
		list[left] = aux
	}
}

func quickSort2(list []int, partition func(list []int) int) {
	if len(list) > 1 {
		p := partitionLeftMost(list)
		quickSort2(list[:p], partition)
		quickSort2(list[p:], partition)
	}
}

func partitionRightMost(list []int) int {
	pivotIndex := len(list) - 1
	pivot, i := list[pivotIndex], 0

	for idx, el := range list {
		if el < pivot {
			list[idx] = list[i]
			list[i] = el
			i += 1
		}
	}

	aux := list[i]
	list[i] = list[pivotIndex]
	list[pivotIndex] = aux

	return i
}

func partitionLeftMost(list []int) int {
	pivotIndex := 0
	pivot, i := list[pivotIndex], len(list)-1

	for j := i; j > pivotIndex; j-- {
		if list[j] > pivot {
			aux := list[j]
			list[j] = list[i]
			list[i] = aux
			i -= 1
		}
	}

	aux := list[i]
	list[i] = list[pivotIndex]
	list[pivotIndex] = aux

	return i + 1
}
